pywr/recorders/calibration.py

```diff
@@ -82,3 +82,84 @@ def values(self):
         return 1.0 - np.sum((obs-mod)**2, axis=0)/np.sum((obs-obs_mean)**2, axis=0)
 
 
+class CoefficientOfDeterminationNodeRecorder(AbstractComparisonNodeRecorder):
+    """ Recorder evaluates the Coefficient of Determination of model and observed """
+    def values(self):
+        mod = self.data        
+        obs = self._aligned_observed
+        n_items = len(mod)    
+        return ((n_items*np.sum(mod*obs, axis=0)-np.sum(mod, axis=0)*np.sum(obs, axis=0))/(np.sqrt((n_items*np.sum(mod**2, axis=0)-(np.sum(mod, axis=0))**2)*(n_items*np.sum(obs**2, axis=0)-(np.sum(obs, axis=0))**2))))**2
+
+
+class AbstractComparisonStorageRecorder(NumpyArrayStorageRecorder):
+    """ Base class for all Recorders performing timeseries comparison of `Node` storage
+
+
+    """
+    def __init__(self, model, node, observed, **kwargs):
+        super(AbstractComparisonStorageRecorder, self).__init__(model, node, **kwargs)
+        self.observed = observed
+        self._aligned_observed = None
+
+    def setup(self):
+        super(AbstractComparisonStorageRecorder, self).setup()
+        # Align the observed data to the model
+        from pywr.parameters import align_and_resample_dataframe
+        self._aligned_observed = align_and_resample_dataframe(self.observed, self.model.timestepper.datetime_index)
+
+
+class RootMeanSquaredErrorStorageRecorder(AbstractComparisonStorageRecorder):
```

Comment:
This is great. I do wonder if we should make both sets of recorders use a shared set of functions to do the metric calculations? That might be more maintainable. I'll think about it.